changefeedccl: warn user when creating changefeed that is a duplicate

log-head opened the pull request:
Previously, although it is possible for changefeeds to be created with the same destination URI, in some cases this is due to an error or typo that can be hard to detect, and can cause issues in the downstream environment. To address this, this patch issues a warning when a changefeed is created for the same tables and same URI. The sinkURI and table IDs are hashed and saved to job_info. CREATE CHANGEFEED statements will look for matching rows. New cluster setting for adding hashes to job_info and raising warnings on dups changefeed.duplicate_check.enabled.
    
The warning looks like:
NOTICE: One or more changefeed jobs are running with the same table(s) and sink URI: job ID [jobID]
    
Fixes: #144262
Epic: CRDB-49296
    
Release note (general change): Provide a warning when creating a changefeed that matches the sink URI and tables of a running changefeed created or resumed after enabling the cluster setting SET CLUSTER SETTING changefeed.duplicate_check.enabled = true. 
Limitation: enabling the cluster setting for the 1st time will not effect changefeeds already running. Changefeeds will need to be resumed, or created for the first time, while the cluster setting is enabled in order to raise the warning on future changefeeds duplicating these.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147713)
<!-- Reviewable:end -->

aerfrei commented on pkg/ccl/changefeedccl/alter_changefeed_stmt.go:
Does this mean we only update the identity if this feature is on?
I'm imagining a scenario like this: 
First the flag is on
Create changefeed into sinkURI1
Turn flag off
Alter the changefeed to target URI2
Flag back on
Create new changefeed into first uri (sinkURI1)

Will we flag this when we shouldn't? Similarly, if the new feed was into URI2, would we be missing that?

log-head commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
That's right. Only changefeeds created/altered while this flag is on are able to be found as duplicates.

aerfrei commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -1899,6 +1941,37 @@ func failureTypeForStartupError(err error) changefeedbase.FailureType {
 	return changefeedbase.OnStartup
 }
 
+func addDuplicateCheckToJobInfo(
```
What does this function do that's different than UpdateChangefeedIdentityHashToJobInfo? The names sound similar and both seem to write the computed hash to the job info table. I'm finding the naming confusing. If it's possible to encapsulate the hash logic in a function I think it might the differences more transparent. Is it that Update takes place inside a transaction?

aerfrei commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
I think it sounds right to me that we expect to "resume" the changefeed job when we create a new changefeed. This is what you saw in local testing, right?

aerfrei commented on pkg/ccl/changefeedccl/alter_changefeed_stmt.go:
Is it fine to call "update" here if the hash doesn't exist (this will call delete). If so, should we just always call update and not have a separate add function?

aerfrei reviewed (commented):
I think this looks good and makes sense to me! I left a few comments

aerfrei commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -11671,3 +11671,63 @@ func TestCloudstorageParallelCompression(t *testing.T) {
+	sqlDB.Exec(t, `CREATE TABLE t4 (a INT PRIMARY KEY, b STRING)`)
+	res = sqlDB.QueryStr(t, `CREATE CHANGEFEED FOR d.t3, d.t4 INTO 'null://'`)
+
+	jobID, err = strconv.Atoi(res[0][0])
```
nit: The [0][0] looks kind of funny. If you use QueryRow, would it go away?

aerfrei commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -299,6 +327,7 @@ func changefeedPlanHook(
 		// changeFrontier.manageProtectedTimestamps for more details on the handling
 		// of protected timestamps.
 		var sj *jobs.StartableJob
+		// sj.InfoStorage()
```
nit: remove

aerfrei commented on pkg/ccl/changefeedccl/alter_changefeed_stmt.go:
```diff
@@ -263,6 +263,12 @@ func alterChangefeedPlanHook(
 		}); err != nil {
 			return err
 		}
+		jobRecord.JobID = jobID
```
Why are we adding the jobID to the jobRecord? It looks like it's initialized with a jobID

aerfrei commented on pkg/jobs/registry.go:
```diff
@@ -694,6 +697,69 @@ func (r *Registry) CreateIfNotExistAdoptableJobWithTxn(
+	if row == nil {
+		return false, nil
+	}
+	return true, nil
```
Even though it's a (very?) low probability event, technically hash collisions are possible. Do we need some kind of additional check here to make sure that the hash collisions corresponds to two rows with the same sink and targets? I think it would be a pretty funny scenario if a customer can't start a changefeed (without turning off this feature) because {sinkURI1, Table1} and {sinkURI2, Table2} map to the same hash

log-head replied:
The probability of a collision is astronomically low - with 256 bits, we won't expect a collision until about 2^128 hashes have been added. Additional checks could be expensive and would have to be done each time a duplicate hash is found. We'd either need to add another info_type row to job info that holds the sinkURI and table IDs (extra storage, and then the comparison could be slow), or unpack the full protobuf of the existing row.

aerfrei replied:
Yeah, ultimately this is probably right. I'll take those odds.
That being said, do we think it would be useful to show in the error message where the collision is (with the jobId of the duplicate feed or the create statement)? It seems like fetching that would be about as difficult as making this check.

asg0451 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -138,6 +144,28 @@ func changefeedTypeCheck(
+		return err
+	}
+	if duplicateJobID != jobspb.InvalidJobID {
+		err = p.SendClientNotice(ctx, pgnotice.Newf("One or more changefeed jobs are running with the same table(s) and sink URI: job ID %d, ...", duplicateJobID), true)
```
i think copilot was right here -- rm ellipsis

log-head commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -11671,3 +11671,69 @@ func TestCloudstorageParallelCompression(t *testing.T) {
+	jobID = min(jobID, expectNoticeGetJobID(t, s.Server, `CREATE CHANGEFEED FOR d.t0 INTO 'null://'`, false, jobspb.InvalidJobID))
+	sqlDB.Exec(t, `RESUME JOB $1`, jobID1)
+	sqlDB.CheckQueryResultsRetry(t, fmt.Sprintf("SELECT count(*) FROM [SHOW JOBS] WHERE job_id = %d and job_type='CHANGEFEED' and status='running'", jobID1), [][]string{{"1"}})
+	time.Sleep(3 * time.Second) // Resuming job - status is set to running before the code for resuming the job runs. So, takes a few seconds before it can be tracked for dups.
```
The jobs.Resumer method is invoked after the job status is set to "running". So there is a time between when status is running and when the changefeed identity hash has been added to the job_info table. This sleep is to give time for the jobs.Resumer function (Resume in changefeed_stmt.go) to execute.

(pkg/jobs/registry.go in stepThroughStateMachine, case StateRunning)

log-head committed:
changefeedccl: warn user when creating changefeed that is a duplicate

Previously, although it is possible for changefeeds to be created with
the same destination URI, in some cases this is due to an error or typo
that can be hard to detect, and can cause issues in the downstream
environment. To address this, this patch issues a warning when a
changefeed is created for the same tables and same URI. The sinkURI and
table IDs are hashed and saved to job_info. CREATE CHANGEFEED statements
will look for matching rows. New cluster setting for adding hashes to
job_info and raising warnings on dups
changefeed.duplicate_check.enabled.

The warning looks like:
NOTICE: One or more changefeed jobs are running with the same table(s)
and sink URI: job ID [jobID]

Fixes: #144262
Epic: CRDB-49296

Release note (general change): Provide a warning when creating a
changefeed that matches the sink URI and tables of a running changefeed
created or resumed after enabling the cluster setting
SET CLUSTER SETTING changefeed.duplicate_check.enabled = true.
Limitation: enabling the cluster setting for the 1st time will not effect
changefeeds already running. Changefeeds will need to be resumed, or
created for the first time, while the cluster setting is enabled in order
to raise the warning on future changefeeds duplicating these.

asg0451 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -138,6 +138,29 @@ func changefeedTypeCheck(
+		return err
+	}
+	if found {
+		err = p.SendClientNotice(ctx, pgnotice.Newf("One or more changefeed jobs are running with the same table(s) and sink URI: job ID %d", duplicateJobID), true)
```
pls update the commit msg to reflect the actual wording of the notice

asg0451 commented on pkg/ccl/changefeedccl/changefeedbase/settings.go:
```diff
@@ -369,3 +369,10 @@ var RetryBackoffReset = settings.RegisterDurationSettingWithExplicitUnit(
 	settings.DurationInRange(1*time.Second, 1*time.Hour),
 )
+
+var FeatureChangefeedDuplicateCheckEnabled = settings.RegisterBoolSetting(
```
pls rename the var and the setting name to be like the other ones, eg `changefeed.duplicate_check.enabled`

asg0451 commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -138,6 +138,29 @@ func changefeedTypeCheck(
+func maybeShowDuplicateChangefeedNotice(
+	ctx context.Context, jr *jobs.Record, txn isql.Txn, p sql.PlanHookState,
+) error {
+	// For purposes of duplicates, an identity is a combination of sinkURI and target table IDs.
+	// The changefeed identity is hashed and added to the job_info table on (1) new changefeed creation,
+	//  (2) when a changefeed job is altered, and (3) when a changefeed job is resumed.
+	// On changefeed creation, active jobs are searched for matching changefeed identity hashes.
```
this comment should be on FindDuplicateChangefeedJob probably

asg0451 commented on pkg/ccl/changefeedccl/alter_changefeed_stmt.go:
i'm not 100% sure of the semantics of p.InternalSQLTxn() -- will this be done transactionally with the above update in all cases?

asg0451 commented on pkg/ccl/changefeedccl/helpers_test.go:
```diff
@@ -910,6 +910,75 @@ func expectNotice(
+// expectNoticeGetJobID is used for checking duplicate changefeed notices when creating changefeeds.
+// If multiple notices returned, it checks all to see that the expected notice is present
+// exactly once.
+func expectNoticeGetJobID(
```
nit: change the name to make it more explicit that this is some super specific helper to do with dupe checking.

asg0451 commented on pkg/jobs/registry.go:
```diff
@@ -688,6 +691,98 @@ func (r *Registry) CreateIfNotExistAdoptableJobWithTxn(
+
+// GetChangefeedIdentityHash computes the hash of a changefeed job's identifying
+// attributes - sink URI and table IDs.
+func (r *Registry) GetChangefeedIdentityHash(
```
nit: "Get" implies a field accessor. can you make the name more clearly indicate it's a calculation?

asg0451 commented on pkg/jobs/registry.go:
```diff
@@ -688,6 +691,98 @@ func (r *Registry) CreateIfNotExistAdoptableJobWithTxn(
+func (r *Registry) GetChangefeedIdentityHash(
+	ctx context.Context, jobDetails jobspb.ChangefeedDetails,
+) ([]byte, error) {
+	// The space of hash values from sha256 that we won't worry about collisions.
```
nit: grammar/meaning

asg0451 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -11808,3 +11808,84 @@ func TestCloudstorageParallelCompression(t *testing.T) {
+	sqlDB := sqlutils.MakeSQLRunner(s.DB)
+	var jobID jobspb.JobID
+
+	t.Run("resume existing changefeed", func(t *testing.T) {
```
pls use the cdcTest framework if possible
